tests/core/test_callable_utils.py

```diff
@@ -0,0 +1,880 @@
+    COMMENT 'Function with List[Dict[str, int]] input'
+    AS $$
+return str(a)
+    $$;
+    """
+    assert sql_body.strip() == expected_sql.strip()
+
+def test_function_with_heavily_nested_structure():
+    def func_with_heavily_nested(a: List[Dict[str, List[Dict[str, int]]]]) -> str:
+        """
+        A function that accepts a heavily nested structure of lists and dictionaries.
+
+        Args:
+            a: A list of dictionaries where the key is a string and the value is a list of dictionaries
+               with string keys and integer values.
+        
+        Returns:
+            str: A string representation of the nested structure
+        """
+        return str(a)
+
+    sql_body = generate_sql_function_body(
+        func_with_heavily_nested,
+        "Function with heavily nested structure List[Dict[str, List[Dict[str, int]]]]",
+        "test_catalog",
+        "test_schema"
+    )
+
+    expected_sql = """
+    CREATE OR REPLACE FUNCTION test_catalog.test_schema.func_with_heavily_nested(a ARRAY<MAP<STRING, ARRAY<MAP<STRING, INTEGER>>>> COMMENT 'A list of dictionaries where the key is a string and the value is a list of dictionaries with string keys and integer values.')
+    RETURNS STRING
+    LANGUAGE PYTHON
+    COMMENT 'Function with heavily nested structure List[Dict[str, List[Dict[str, int]]]]'
+    AS $$
+return str(a)
+    $$;
+    """
+    
+    assert sql_body.strip() == expected_sql.strip()
+
+def test_function_with_extra_docstring_params_ignored():
+    def func_with_extra_param_in_docstring(a: int) -> str:
+        """
+        A function with extra parameter in docstring.
+
+        Args:
+            a: The first argument
+            b: An extra parameter not in function signature
+        
+        Returns:
+            str: The string representation of the first argument
+        """
+        return str(a)
+
+    # We expect the generated SQL to ignore 'b' since it's not in the function signature
+    sql_body = generate_sql_function_body(
+        func_with_extra_param_in_docstring,
+        "Function with extra docstring parameter",
+        "test_catalog",
+        "test_schema"
+    )
+    
+    expected_sql = """
+    CREATE OR REPLACE FUNCTION test_catalog.test_schema.func_with_extra_param_in_docstring(a INTEGER COMMENT 'The first argument')
+    RETURNS STRING
+    LANGUAGE PYTHON
+    COMMENT 'Function with extra docstring parameter'
+    AS $$
+return str(a)
+    $$;
+    """
+    
+    assert sql_body.strip() == expected_sql.strip()
+
+
+def test_function_with_plain_list_type():
+    def func_with_plain_list_type(a: List) -> str:
+        """
+        A function with a plain List as a parameter type.
+
+        Args:
+            a: A plain list without inner types
+
+        Returns:
+            str: A string representation of the list
+        """
+        return str(a)
+
+    with pytest.raises(ValueError, match="Error in parameter 'a': type typing.List is not supported."):
```

Comment:
Can we make this error message more actionable?

Reply:
yeah, that message is terrible. I refactored the error handling to provide guidance on collections since this was so vague and confusing (required looking at the stack trace, which is NOT good)